### PR TITLE
Feature/layers wave2

### PR DESCRIPTION
**Pull request by SebastianSchildt**
This allows arbitrary attributes. JSON & YAML can generate them (default: The ones from whitelist, alternatives: None (same as now whitelist given) or all (only possible if --strict is not used, otherwise, parsing aborts on non whitelisted attributes).

Confused? Check help:

```
usage: vspec2x.py [-h] [-I dir] [-e EXTENDED_ATTRIBUTES] [-s] [--abort-on-non-core-attribute] [--abort-on-name-style] [--format format] [--no-uuid]
                  [-o overlays] [--json-no-extended-attributes] [--json-all-extended-attributes] [--json-pretty] [--yaml-no-extended-attributes]
                  [--yaml-all-extended-attributes] [-v version] [--all-idl-features] [--gqlfield GQLFIELD GQLFIELD]
                  <vspec_file> <output_file>

Convert vspec to other formats.

positional arguments:
  <vspec_file>          The vehicle specification file to convert.
  <output_file>         The file to write output to.

optional arguments:
  -h, --help            show this help message and exit
  -I dir, --include-dir dir
                        Add include directory to search for included vspec files.
  -e EXTENDED_ATTRIBUTES, --extended-attributes EXTENDED_ATTRIBUTES
                        Whitelisted extended attributes as comma sperated list. Note, that not all exporters will support (all) extended attributes.
  -s, --strict          Use strict checking: Terminate when anything not covered or not recommended by the core VSS specs is found.
  --abort-on-non-core-attribute
                        Terminate when non-core attribute is found.
  --abort-on-name-style
                        Terminate naming style not follows recommendations.
  --format format       Output format, choose one from ['json', 'csv', 'yaml', 'binary', 'franca', 'idl', 'graphql']. If omitted we try to guess form
                        output_file suffix.
  --no-uuid             Exclude uuids from generated files.
  -o overlays, --overlays overlays
                        Add overlays that will be layered on top of the VSS file in the order they appear.

JSON arguments:

  --json-no-extended-attributes
                        Generate no extended attributes.
  --json-all-extended-attributes
                        Generate all extended attributes found in the model (default is generating only those given by the -e/--extended-attributes
                        parameter).
  --json-pretty         Pretty print JSON output.

CSV arguments:
  The CSV exporter does not currently support the no-uuid option.

YAML arguments:

  --yaml-no-extended-attributes
                        Generate no extended attributes.
  --yaml-all-extended-attributes
                        Generate all extended attributes found in the model (default is generating only those given by the -e/--extended-attributes
                        parameter).

BINARY arguments:
  The binary exporter does not support any additional arguments.

FRANCA arguments:
  The Franca exporter does not currently support the no-uuid option.

  -v version            Add version information to franca file.

IDL arguments:
  The DDS-IDL exporter

  --all-idl-features    Generate all features based on DDS IDL 4.2 specification

GRAPHQL arguments:
  The graphql exporter does not currently support the no-uuid option.

  --gqlfield GQLFIELD GQLFIELD
                        Add additional fields to the nodes in the graphql schema. use: <field_name> <description>
```

Overlay to play

```yaml
#
# Example overlay
#
Vehicle:
  type: branch



Vehicle.RandomSpeed:
  datatype: float
  type: sensor
  quality: Average
  unit: km/h
  description: Average speed for the current trip.
```

Potential cmdlines

```
python vspec2json.py ../spec/VehicleSignalSpecification.vspec  -e quality -o test.vspec  --json-pretty out.json
```
Generates quality attribute

```
python vspec2json.py ../spec/VehicleSignalSpecification.vspec  -o test.vspec  --json-pretty out.json
```
Does not generate quality attribute

```
python vspec2json.py ../spec/VehicleSignalSpecification.vspec  -o test.vspec  --json-all-extended-attributes --json-pretty out.json
```
Generates quality attribute

```
python vspec2json.py ../spec/VehicleSignalSpecification.vspec  -o test.vspec  --json-all-extended-attributes --json-pretty --strict out.json
```
Parsing fails

```
python vspec2json.py ../spec/VehicleSignalSpecification.vspec  -e quality -o test.vspec  --json-all-extended-attributes --json-pretty --strict out.json
```
Generates quality attribute

### CODE REVIEW

**Comment by erikbosch** [comment]
Would not the comment above (with the example) be a quite good candidate to add to a markdown documentation file. Then we can discuss if we want layering to be considered as "tooling"-specific and be documented in this repo, or if it shall be part of the overall generated documentation for vss.

File: vspec/model/vsstree.py [hunk 1]
```diff
@@ -36,7 +36,14 @@ class VSSNode(Node):
     comment = ""
     uuid = None
     type: VSSType
-    data_type: VSSDataType
+    datatype: VSSDataType
+
+    core_attributes = ["type", "children", "datatype", "description", "unit", "uuid", "min", "max", "allowed",
+                            "aggregate", "default" , "instances", "deprecation", "arraysize", "comment", "$file_name$"]
+
+    # List of accepted extended attributes. In strict terminate if an attribtute is
```
Comment (erikbosch) [inline vspec/model/vsstree.py:44]: Minor spelling error "attribtute"

File: vspec2x.py [hunk 2]
```diff
@@ -53,6 +53,8 @@ def from_string(s):
 def main(arguments):
     parser.add_argument('-I', '--include-dir', action='append',  metavar='dir', type=str,  default=[],
                         help='Add include directory to search for included vspec files.')
+    parser.add_argument('-e', '--extended-attributes', type=str,  default="",
+                        help='Whitelisted extended attributes as comma sperated list. Note, that not all exporters will support (all) extended attributes.')
     parser.add_argument('-s', '--strict', action='store_true',
                         help='Use strict checking: Terminate when anything not covered or not recommended by the core VSS specs is found.')
```
Comment (erikbosch) [inline vspec2x.py:59]: In some way we should rephrase this as the term "core VSS specs" is not true any longer as we consider "extensions". 

Maybe something like:

"Terminate when anything not covered or not recommended by VSS language or extensions is found""

File: vspec/model/vsstree.py [hunk 3]
```diff
@@ -72,53 +79,39 @@ def __init__(self, name, source_dict: dict, parent=None, children=None, break_on
         super().__init__(name, parent, children)
         try:
             VSSNode.validate_vss_element(source_dict, name)
-        except NonCoreAttributeException as e:
+        except UnknownAttributeException as e:
             print("Warning: {}".format(e))
             if break_on_noncore_attribute:
```
Comment (erikbosch) [inline vspec/model/vsstree.py:84]: Replace "noncore" with "unknown"?

**Review by danielwilms — COMMENTED** [review]
I tried it and what I tried works. For me an open question is where to place the arguments. But  it's a really nice feature!

File: vssexporters/vss2json.py [hunk 4]
```diff
@@ -12,17 +12,22 @@
...
 from vspec.model.vsstree import VSSNode, VSSType
 
-def add_arguments(parser: argparse.ArgumentParser):
-    parser.add_argument('--json-pretty', action='store_true', help=" Pretty print JSON output.")
 
+def add_arguments(parser: argparse.ArgumentParser):
+    parser.add_argument('--json-no-extended-attributes',
+                        action='store_true', help="Generate no extended attributes.")
+    parser.add_argument('--json-all-extended-attributes', action='store_true',
+                        help="Generate all extended attributes found in the model (default is generating only those given by the -e/--extended-attributes parameter).")
```
[the remark this answers is not in the document]
Comment (SebastianSchildt) [reply vssexporters/vss2json.py:20]: 1. I think we do need them for each exporter because not all arguments do make sense for all exporters (i.e. generate "all" is quite useful for json and yaml, but not useful with CSV maybe, at least not easy, because it requires a two-pass iteration over the data model.
2. Looking at it, I agree the "no-extended" seems redundant: With "-e"=Please use this list, with "-all": Generate everyting you can. With -e unset it is basically none

**Comment by SebastianSchildt** [comment]
Addressed smaller comments, and wrote documentation vor vspec2x, not specifically for overlays (I think the workings of that maybe need to go to VSS doc, or maybe a seperate example page here).
"Readable" doc: https://github.com/boschglobal/vss-tools/blob/feature/layers-wave2/docs/vspec2x.md

File: docs/vspec2x.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+## CSV exporter notes
+The CSV exporter does not currently support the no-uuid option.
+
+## FRANCA exporter notes
+The Franca exporter does not currently support the no-uuid option.
+
+## DDS-IDL exporter notes
+
+### --all-idl-features
+Will also generate non-payload const attributes such as unit/datatype. Default is not to generate them/comment them out becasue at least Cyclone DDS and FastDDS do not support const. For more information check the [DDS-IDL exporter docs](VSS2DDSIDL.md).
```
Comment (erikbosch) [inline docs/vspec2x.md:132]: Spelling error "becasue"

File: docs/vspec2x.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+## GRAPHQL exporter notes
+
+### --gqlfield GQLFIELD GQLFIELD
+Add additional fields to the nodes in the graphql schema. use: <field_name> <description>
+
+
+## Writing yor own exporter
```
Comment (slawr) [inline docs/vspec2x.md:140]: spelling: yor

**Comment by SebastianSchildt** [comment]
Current version fixes the problem, that data was lost when overriding elements that will be instantiated (this was observed here https://github.com/COVESA/vss-tools/pull/168 )

This has been fixed. Please double check it still generates the same output as current master (It should, but this includes changes to the merge/deepcopy code path....)

Also it detects a little more "stupidity" in vspec and terminates more gracefully (specifically declaring a sensor/attribute/actuator without datatype or trying to change the branch type to sensor/actuator/attribute or vice versa in an overly).

Typo fixes still pending, but keep looking :)

**Comment by SebastianSchildt** [comment]
Please check @danielwilms

**Comment by danielwilms** [comment]
works for me, didn't go through the details, but should be fine to merge. Everything else is a bug fix.

**Review by danielwilms — APPROVED** [review]
looks great
